tenpy/algorithms/mps_compress.py

```diff
@@ -0,0 +1,345 @@
+r"""Compression of a MPS.
+
+
+"""
+
+# Copyright 2019 TeNPy Developers, GNU GPLv3
+
+import numpy as np
+
+from ..linalg import np_conserved as npc
+from .truncation  import svd_theta
+from ..networks import mps, mpo
+
+__all__ = [
+    'make_U', 'make_UI', 'make_UII', 'make_WII', 'mps_compress', 'svd_two_site', 'apply_mpo'
+]
+
+def make_U(H, dt, which='II'):
+    r"""Creates the UI or UII propagator for a given Hamiltonian.
+
+    Parameters
+    ----------
+    H : :class:`~tenpy.networks.mpo.MPO`
+    dt : float|complex
+
+    Returns
+    -------
+    U : :class:`~tepy.networks.mpo.MPO`
+        The propagator.
+    """
+    if which=='II':
+        return make_UII(H, dt)
+    if which=='I':
+        return make_UI(H, dt)
+    raise ValueError("'%s' not implemented"%which)
+
+def make_UI(H, dt):
+    r"""Creates the UI propagator for a given Hamiltonian.
+
+    Parameters
+    ----------
+    H : :class:`~tenpy.networks.mpo.MPO`
+        The Hamiltonian to use
+    dt : float|complex
+        the time step per application of the propagator
+
+    Returns
+    -------
+    UI : :class:`~tenpy.networks.mpo.MPO`
+        The propagator.
+
+    """
+
+    U = [H.get_W(i).astype(np.result_type(dt,H.dtype)).transpose(['wL', 'wR', 'p', 'p*']) for i in range(H.L)]
+
+    IdLR = []
+    for i in range(0, H.L):  # correct?
+        U1 = U[i]
+        U2 = U[(i+1) % H.L]
+        IdL = H.IdL[i+1]
+        IdR = H.IdR[i+1]
+        assert IdL is not None and IdR is not None
+        U1[:, IdL, :, :] = U1[:, IdL, :, :] + dt * U1[:, IdR, :, :]
+        keep=np.ones(U1.shape[1], dtype=bool)
+        keep[IdR] = False
+        U1.iproject(keep, 1)
+        if H.finite and i + 1 == H.L:
+            keep = np.ones(U2.shape[0], dtype=bool)
+            assert H.IdR[0] is not None
+            keep[H.IdR[0]] = False
+        U2.iproject(keep, 0)
+
+        if IdL > IdR:
+            IdLR.append(IdL - 1)
+        else:
+            IdLR.append(IdL)
+
+    IdL = H.IdL[0]
+    IdR = H.IdR[0]
+    assert IdL is not None and IdR is not None
+    if IdL > IdR:
+        IdLR_0 = IdL - 1
+    else:
+        IdLR_0 = IdL
+    IdLR = [IdLR_0] + IdLR
+
+    return mpo.MPO(H.sites, U, H.bc, IdLR, IdLR, np.inf)
+
+
+def make_UII(H, dt):
+    r"""Creates the UII propagator for a given Hamiltonian.
+
+    Parameters
+    ----------
+    H : :class:`~tenpy.networks.mpo.MPO`
+        The Hamiltonian to use
+    dt : float|complex
+        the time step per application of the propagator
+
+    Returns
+    -------
+    UII : :class:`~tenpy.networks.mpo.MPO`
+        The propagator.
+
+    """
+    dtype = np.result_type(dt, H.dtype)
+    IdL=H.IdL
+    IdR=H.IdR
+    v = np.empty_like(IdL)
+
+    c0 = np.zeros((1, H[0].num_q), dtype=np.int) #TODO 
+    U = []
+    for i in xrange(0, self.L):
+            W = H.get_W(i).itranspose(['wL','wR', 'p', 'p*'])
+            W = W.to_ndarray()
+            aL = range(W.shape[0]); aL.remove(IdL[i]); aL.remove(IdR[i]);

```

Comment:
Might be more intuitive and actually faster to use boolean arrays:
```
aL = np.ones(W.shape[0], dtype=np.bool)
aL[IdL[i]] = aL[IdR[i] = False
```
Also, it might make more sense to name it `proj_L` instead of `aL`.
Same for `aR`, of course.